numpy/_core/src/umath/stringdtype_ufuncs.c

```diff
@@ -0,0 +1,1410 @@
+        char *out = data[2];                                                  \
+        npy_intp in1_stride = strides[0];                                     \
+        npy_intp in2_stride = strides[1];                                     \
+        npy_intp out_stride = strides[2];                                     \
+                                                                              \
+        return multiply_loop_core_##shortname(                                \
+                N, in1, in2, out, in1_stride, in2_stride, out_stride,         \
+                has_null, has_nan_na, has_string_na, default_string, idescr,  \
+                odescr);                                                      \
+    }                                                                         \
+                                                                              \
+    static int multiply_left_##shortname##_strided_loop(                      \
+            PyArrayMethod_Context *context, char *const data[],               \
+            npy_intp const dimensions[], npy_intp const strides[],            \
+            NpyAuxData *NPY_UNUSED(auxdata))                                  \
+    {                                                                         \
+        StringDTypeObject *idescr =                                           \
+                (StringDTypeObject *)context->descriptors[1];                 \
+        StringDTypeObject *odescr =                                           \
+                (StringDTypeObject *)context->descriptors[2];                 \
+        int has_null = idescr->na_object != NULL;                             \
+        int has_nan_na = idescr->has_nan_na;                                  \
+        int has_string_na = idescr->has_string_na;                            \
+        const npy_static_string *default_string = &idescr->default_string;    \
+        npy_intp N = dimensions[0];                                           \
+        char *in1 = data[0];                                                  \
+        char *in2 = data[1];                                                  \
+        char *out = data[2];                                                  \
+        npy_intp in1_stride = strides[0];                                     \
+        npy_intp in2_stride = strides[1];                                     \
+        npy_intp out_stride = strides[2];                                     \
+                                                                              \
+        return multiply_loop_core_##shortname(                                \
+                N, in2, in1, out, in2_stride, in1_stride, out_stride,         \
+                has_null, has_nan_na, has_string_na, default_string, idescr,  \
+                odescr);                                                      \
+    }
```

Comment:
I wish we didn't need these giant macros, but don't want to move to c++ just for that here now.

Reply:
Yeah, that's where I'm at with this as well.